### ISSUE TITLE
[accessibility] Accessibility support

### ISSUE THREAD

**Issue by bstruthers** [ceolter/ag-grid#19]
Adding ARIA roles such as 
- [grid](http://www.w3.org/TR/wai-aria/roles#grid)
- [row](http://www.w3.org/TR/wai-aria/roles#row)
- [columnheader](http://www.w3.org/TR/wai-aria/roles#columnheader)
- [gridcell](http://www.w3.org/TR/wai-aria/roles#gridcell)

would improve accessibility for the data since it's not being rendered with a table.

**Comment by amcdnl** [comment]
ng-aria would be easy to do this...

**Comment by bstruthers** [comment]
My understanding of ng-aria is that it adds the attributes to the directives it knows (ngModel, ngDisabled, ngShow, ngHide, ngClick, ngDblClick, and ngMessages).  How would ng-aria know that angular-grid's div is actually a columnheader etc?

**Comment by ceolter** [comment]
Hi @bstruthers, I'll be honest, I wasn't thinking of accessibility, I was (still am!!) busy getting the basic requirements in, such as paging and cell selection. Thanks for raising this with me, I'll stick it on my list.

**Comment by tommut** [comment]
Just wanted to add my encouragement for this. The story of angular-grid really resonated with me - my team and I have been using jqWidgets jqxGrid, but due to poor premium support and many required workarounds due to their not playing well in the angular lifecycle, we made the switch to Angular ui-grid.  We started switching over to it, but the one thing that killed it was when someone noticed that accessibility/508c compliance was not available, and their github issue on the topic said they had no immediate plans for it.  So back to jqxGrid we went.  Your project is great, and I would love if I could show the corporate overlords that the Accessibility checkbox can be checked...   I think the main compliance areas we are looking at are the the aria-role attributes added to the grid tags, and keyboard support for navigating the grid and activating filter dialog, etc.  I know Accessibility is not the sexiest feature to work on, but once the grid basics are implemented if you get a chance to throw some love to enabling accessibility I'd buy you a beer ;)

**Comment by hohwille** [comment]
To be honest: The web is pain. To really address accessibility you do not only need some ARIA attributes. You need keyboard control. However HTML, CSS and JS is the poorest technology on earth for fundamental things. So you need to handle keyboard events for cursor key strokes and handle the focus inside the grid yourself. That is a hole lot of work. It would be great if the projects providing grids could merge their energy to finally provide something addressing all requirements and demands for end-users. Maybe you can at least get inspiration and code hints from other grid projects to get this done. After all one would expect the W3C to specify a grid including JS API and make the browser vendors provide an accessible and device optimized view for it. This is such a fundamental requirement that was already solved 15 years ago in every native UI framework such as Swing, etc.
Now we have year 2015 and the web is still stoneage while you can do cool 3D graphics in the browser easily. Good luck mate!

**Comment by ceolter** [comment]
> It would be great if the projects providing grids could merge their energy to finally provide something addressing all requirements and demands for end-users.

now that would be nice!! but the problem is, with the exception of @amcdnl's grid and slick grid (two different grids), i've not been that impressed with the other grids, so i don't think getting everyone to work together would provide the genius you are hoping for :(

**Comment by ceolter** [comment]
tracked by https://ag-grid.atlassian.net/browse/AG-38

**Comment by stevefaulkner** [comment]
@ceolter what is the status of ARIA implementation support? We have a client who is using your grid component, but it does not expose grid semantics, so users of assistive technology cannot make sense of the UI. Would appreciate an idea of whether support is in development and if so, what is the timeline for release. This will allow us to advise the client as to whether they will need to seek an accessible alternative or can wait until support is implemented.

**Comment by arunsark** [comment]
@ceolter what is the status of ARIA implementation support? we are evaluating to buy enterprise license and WCAG-Level A is must for us. Is this something which you are planning to address.

**Comment by makinggoodsoftware** [comment]
Hi,

Just to let you know that as of 11.0.0 this has been implemented, see the changelog for more details..

https://www.ag-grid.com/change-log/changeLogIndex.php#gsc.tab=0
